carsus/io/output/base.py

```diff
@@ -240,188 +260,222 @@ def calculate_collisional_strength(row, temperatures,
+
         """
 
-        gf_levels = self.gfall_reader.levels.reset_index()
-        gf_levels['source'] = 'gfall'
+        logger.info('Ingesting energy levels.')
+        gf_levels = self.gfall_reader.levels
+        gf_levels['ds_id'] = 2
 
-        if len(self.chianti_ions) > 0:
-            logger.info('Ingesting levels from Chianti.')
-            ch_levels = self.chianti_reader.levels.reset_index()
-            ch_levels['source'] = 'chianti'
+        if self.chianti_reader is not None:
+            ch_levels = self.chianti_reader.levels
+            ch_levels['ds_id'] = 4
         else:
             ch_levels = pd.DataFrame(columns=gf_levels.columns)
 
-        levels = pd.concat([gf_levels, ch_levels], sort=True)
+        if self.cmfgen_reader is not None:
+            cf_levels = self.cmfgen_reader.levels
+            cf_levels['ds_id'] = 5
+        else:
+            cf_levels = pd.DataFrame(columns=gf_levels.columns)
+
+        levels = pd.concat([gf_levels, ch_levels, cf_levels], sort=True)
         levels['g'] = 2*levels['j'] + 1
         levels['g'] = levels['g'].astype(np.int)
         levels = levels.drop(columns=['j', 'label', 'method'])
-        levels = levels.reset_index(drop=True)
+        levels = levels.reset_index()
         levels = levels.rename(columns={'ion_charge': 'ion_number'})
-        levels = levels[['atomic_number',
-                         'ion_number', 'g', 'energy', 'source']]
+        levels = levels[['atomic_number', 'ion_number', 'g', 'energy', 
+                         'ds_id', 'priority']]
 
         levels['energy'] = levels['energy'].apply(lambda x: x*u.Unit('cm-1'))
         levels['energy'] = levels['energy'].apply(
             lambda x: x.to(u.eV, equivalencies=u.spectral()))
         levels['energy'] = levels['energy'].apply(lambda x: x.value)
 
+        # Solve priorities and set attributes for later use.
+        self.gfall_ions, self.chianti_ions, self.cmfgen_ions = self.solve_priorities(levels)
+
+        to_string = lambda x: [f"{convert_atomic_number2symbol(ion[0])} {ion[1]}" \
+                                    for ion in sorted(x)]
+
+        gfall_str = ', '.join(to_string(self.gfall_ions))
+        logger.info(f'GFALL selected species: {gfall_str}.')
+        
+        if len(self.chianti_ions) > 0:
+            chianti_str = ', '.join(to_string(self.chianti_ions))
+            logger.info(f'Chianti selected species: {chianti_str}.')
+        
+        if len(self.cmfgen_ions) > 0: 
+            cmfgen_str = ', '.join(to_string(self.cmfgen_ions))
+            logger.info(f'CMFGEN selected species: {cmfgen_str}.')
+
+        # Concatenate ground levels from NIST
         ground_levels = self.ground_levels
-        ground_levels.rename(
-            columns={'ion_charge': 'ion_number'}, inplace=True)
-        ground_levels['source'] = 'nist'
+        ground_levels = ground_levels.rename(columns={'ion_charge': 'ion_number'})
+        ground_levels['ds_id'] = 1
```

Comment:
what is a `ds_id`?

Reply:
that is `data_source_id` - let's use this.

Reply:
I'll open an issue.